Seconday Menu Indentation System

Allcharles opened the pull request:
Improved secondary menu indentation system so that indentation does not need to be defined by page components. Components specify their parent, which is then used to calculate its indentation level and organise sub-links.

Edit 1:
Order and indentation can now be calculated for all sub-links automatically. Order for top level links must still be set manually.

Closes #58

Allcharles committed:
Replaced indentation system

Allcharles committed:
Working menu unit tests

Secondary menu component unit test fail on firefox because function is not supported

Allcharles committed:
Firefox compatible unit tests

atruskie commented on src/app/interfaces/menusInterfaces.ts:
```diff
@@ -53,12 +53,12 @@ export interface Order {
+   * Indentation of link. Value is calculated by SecondaryMenuComponent
    */
-  indentation: number;
+  indentation?: number;
```
if this option is no longer user-configurable  then let's remove it from the interface

Allcharles replied:
I am unsure how to remove this, there is no way i can pass the indentation value of the menu item outside of the values specified in the interface. The only option i have is to turn all of the menu item types into classes which would be a bit of a rewrite.

atruskie commented on src/app/component/shared/secondary-menu/secondary-menu.component.ts:
```diff
@@ -50,8 +51,26 @@ export class SecondaryMenuComponent implements OnInit, OnDestroy {
         while (menuRoute.parent) {
           menuRoute = menuRoute.parent;
           parentMenuRoutes.push(menuRoute);
+          indentation++;
```
Hmmmm... recursive loop

atruskie commented on src/app/component/shared/secondary-menu/secondary-menu.component.ts:
```diff
@@ -50,8 +51,26 @@ export class SecondaryMenuComponent implements OnInit, OnDestroy {
+          current.order = { indentation };
+        }
+
+        parentMenuRoutes.forEach(parent => {
```
it isn't readily obvious why you have to set indentation on parent routes?

This loop has the potential to run many times...

atruskie reviewed (changes requested):
Okay. So the original reason that I very intentionally set the indentation to a user-controlled integer is that it is simple. The menu indentation is something that very rarely changes, and as an integer is super easy to fix. More than all of that, it is fast!

Making the indentation value dynamic means there are fewer chances for the value to be incorrect, so I'm willing to entertain this idea.

However, the current implementation has worse theoretical performance; it went from O(1) to O(2N).

So here's the solution I'd like you to try and get to:

When a menu item is created, it directly keeps a reference to its parent. When that happens it gets the parent's indentation level, adds one and stores the indentation value on itself. 
In this way, lookup of indentation values is O(1) (which is good because lookup happens many times).
The flaw here is if we need to change indentation value while the app is running - but that is not a requirement we have yet, and if we do need to we can recompute the chain once, rather than on every render.

Allcharles committed:
Updated order and indentation system

Changed it so that order and indentation can be calculated by a menu items parent.

If it does not have a parent, order is set by the items specified order.

If it does not have a prent, indentation is set to 0.

Allcharles committed:
Fixed unit tests
